### PR TITLE
New_solution_for_persistent_forms

### PR DESCRIPTION
**Pull request by lilitkarapetyan**
## 🔨Work carried out:

- [x] Tests pass

## Confirmations

- [ ] I have chosen reviewers for my PR.
- [ ] I have assigned myself to this PR.
- [ ] I have chosen an appropriate label for the PR.
- [ ] I have completed the mandatory sections of this document.
- [ ] I have deleted any unused sections.
- [ ] I confirm that I have checked for required README updates and acted accordingly.

### CODE REVIEW

**Comment by IanMayo** [comment]
Thanks Lilit.  Data isn't lost, but we still have the flicker.

Could you investigate an alternate solution please?

I'm thinking of something like this:
* we only re-create the JSON editor if some test is passed, such as a new item is being edited

I'm happy to tidy this PR up, and merge it - but the users I spoke to last week are expecting a solution that doesn't involve flicker.  This is what they are used to: when they are instant-messaging with their family, the instant messenger doesn't flicker when a new message is received.  So, they know it is fixable, and so find the flicker annoying.

File: packages/components/src/local/molecules/json-editor/index.tsx [hunk 1]
```diff
@@ -74,19 +73,25 @@ export const JsonEditor: React.FC<Props> = ({
...
+
   const OnSave = () => {
     saveMessage && saveMessage()
     expiredStorage.removeItem(genLocalStorageId())
+    expiredStorage.removeItem(getEditLocalStorgaeid())
   }
 
   const onPopupCancel = (): void => {
-    expiredStorage.removeItem(genLocalStorageId())
+    // removePlanning
```
Comment (IanMayo) [inline packages/components/src/local/molecules/json-editor/index.tsx:71]: @lilitkarapetyan  - do you remember what the `removePlanning` comment is for?

File: packages/components/src/local/molecules/json-editor/index.tsx [hunk 2]
```diff
@@ -74,19 +73,25 @@ export const JsonEditor: React.FC<Props> = ({
     return memoryName
   }
 
+  const getEditLocalStorgaeid = (): string => `edit-${genLocalStorageId()}`
```
Comment (IanMayo) [inline packages/components/src/local/molecules/json-editor/index.tsx]: @lilitkarapetyan - we have two methods with similar names (`getEditLocalStorageId` and `genLocalStorageId`).  Can you put in comments to indicate what they are both for, please?

File: packages/components/src/local/molecules/json-editor/index.tsx [hunk 3]
```diff
@@ -230,12 +235,14 @@ export const JsonEditor: React.FC<Props> = ({
   }, [disableArrayToolsWithEditor && disabled])
 
   useLayoutEffect(() => {
```
Comment (IanMayo) [inline packages/components/src/local/molecules/json-editor/index.tsx:202]: @lilitkarapetyan - I tried to understand what was happening in this method before, and couldn't get it.

My favourite outcome would be to simplify the API/props for the `json-editor`.  I feel we've got some props that do the same job as each other.  If we can't do that, I'll need better comments on the props, and in the method please.

File: packages/components/src/local/molecules/json-editor/index.tsx [hunk 4]
```diff
@@ -263,7 +270,10 @@ export const JsonEditor: React.FC<Props> = ({
                 : null
             }
           </>
-          : !disabled ? <Button color='secondary' onClick={() => { setBeingEdited(true) }} icon='edit'>Edit</Button>
```
Comment (IanMayo) [inline packages/components/src/local/molecules/json-editor/index.tsx:270]: Oh dear, `JSON.stringify('edit')` is too complex for me. Won't `JSON.stringify('edit')` just return `"edit"` ?

**Comment by lilitkarapetyan** [comment]
> Thanks Lilit. Data isn't lost, but we still have the flicker.
> 
> Could you investigate an alternate solution please?
> 
> I'm thinking of something like this:
> 
>     * we only re-create the JSON editor if some test is passed, such as a new item is being edited
> 
> 
> I'm happy to tidy this PR up, and merge it - but the users I spoke to last week are expecting a solution that doesn't involve flicker. This is what they are used to: when they are instant-messaging with their family, the instant messenger doesn't flicker when a new message is received. So, they know it is fixable, and so find the flicker annoying.

Hello Ian, I don't think the problem is with jsonEditor because by default when we don't have anything created with jsonEditor it flashes again when we get a new message, I will work on that and try to fix the flashing.

File: packages/components/src/local/molecules/json-editor/index.tsx (hunk 1 above)
Comment (lilitkarapetyan) [reply packages/components/src/local/molecules/json-editor/index.tsx:71]: Should we delete the chached message ?

File: packages/components/src/local/molecules/json-editor/index.tsx (hunk 2 above)
Comment (lilitkarapetyan) [reply packages/components/src/local/molecules/json-editor/index.tsx]: Ok I will add a comment in the code.

File: packages/components/src/local/molecules/json-editor/index.tsx (hunk 3 above)
Comment (lilitkarapetyan) [reply packages/components/src/local/molecules/json-editor/index.tsx:202]: I will look into it.

Comment (lilitkarapetyan) [reply packages/components/src/local/molecules/json-editor/index.tsx:202]: I don't remember where we use it either, I guess we can remove it from the `jsonEditor`

**Comment by lilitkarapetyan** [comment]
I will try improve it so that when updating a message the scrollbar is in the edit area

File: packages/components/src/local/organisms/planning-messages-list/index.tsx [hunk 5]
```diff
@@ -54,7 +54,7 @@ export const PlanningMessagesList: React.FC<PropTypes> = ({
     }
   }
 
-  useEffect(() => {
+  useLayoutEffect(() => {
```
Comment (IanMayo) [inline packages/components/src/local/organisms/planning-messages-list/index.tsx:53]: Hey @lilitkarapetyan - THIS WORKS PERFECTLY!

Can you explain how it fixes it, please?

**Comment by lilitkarapetyan** [comment]
I'm still working on this PR

**Comment by IanMayo** [comment]
Hello Lilit, thanks for keeping this moving.  But, as I said last week, me preferred outcome is to identify a solution that avoids cache-ing data on table updates.  My preferred outcome is that we use state to prevent JSON-Editor (or its react parents) from being refreshed.

**Comment by lilitkarapetyan** [comment]
> Hello Lilit, thanks for keeping this moving. But, as I said last week, me preferred outcome is to identify a solution that avoids cache-ing data on table updates. My preferred outcome is that we use state to prevent JSON-Editor (or its react parents) from being refreshed.

Hello Ian, okay, I will work on it now

File: packages/components/src/local/organisms/planning-messages-list/index.tsx [hunk 6]
```diff
@@ -4,45 +4,70 @@ import { MessageDetails, MessagePlanning, PerForcePlanningActivitySet, PlannedAc
...
   !7 && console.log('planning selectedOrders: ', selectedOrders, !!setSelectedOrders, messages.length)
 
   const [myMessages, setMyMessages] = useState<MessagePlanning[]>([])
+  const roles: string[] = []
   useEffect(() => {
     const showOrdersForAllRoles = !onlyShowMyOrders
     const myForceMessages = messages.filter((message: MessagePlanning) => isUmpire || message.details.from.forceId === playerForceId)
     const myRoleMessages = myForceMessages.filter((message: MessagePlanning) => showOrdersForAllRoles || message.details.from.roleId === playerRoleId)
-    setMyMessages(myRoleMessages)
+
```
Comment (IanMayo) [inline packages/components/src/local/organisms/planning-messages-list/index.tsx]: Hey @lilitkarapetyan - is this next block where the new magic happens ?

**Comment by IanMayo** [comment]
Hello @lilitkarapetyan - could you update me on where we are with this issue please?

**Comment by lilitkarapetyan** [comment]
> Hello @lilitkarapetyan - could you update me on where we are with this issue please?

Hello Ian, I'm going to graduate

**Comment by lilitkarapetyan** [comment]
I finished it

**Comment by IanMayo** [comment]
Hello @lilitkarapetyan - I'm keen to merge this (hooray!), but I'm just doing a bit of last-minute testing.  We only use `cacheMessage` in one place, `<MessageCreator>`. This component is used in creating new `Orders` messages in the `Planning` tab.  I've tried changing the value of `cacheMessage` to false in that component - and things still seem to work. 

I've pushed the change. Would you mind double-checking?  

That's because if we don't need to use `cacheMessage={true}` anywhere, we can remove a lot of the `expiredStorage` code.

**Comment by IanMayo** [comment]
Aah, I've found an issue @lilitkarapetyan 

When I ask to filter for "Orders" that belong to me, the list doesn't update, but when I cancel the filter, the table grows by one (as it re-processes the first element).  I don't think it matters if we lose live edits when a filter is applied/removed, since the user is deliberately changing the contents of the page.
https://i.gyazo.com/40eb053789a671fabd3058f4adfdfe67.mp4

**Comment by lilitkarapetyan** [comment]
> Hello @lilitkarapetyan - I'm keen to merge this (hooray!), but I'm just doing a bit of last-minute testing. We only use `cacheMessage` in one place, `<MessageCreator>`. This component is used in creating new `Orders` messages in the `Planning` tab. I've tried changing the value of `cacheMessage` to false in that component - and things still seem to work.
> 
> I've pushed the change. Would you mind double-checking?
> 
> That's because if we don't need to use `cacheMessage={true}` anywhere, we can remove a lot of the `expiredStorage` code.

Hello Ian, I will into it now

**Comment by lilitkarapetyan** [comment]
Hello Ian, I will into it now

**Comment by IanMayo** [comment]
I'm not sure how this problem arose, but we have a layout issue with the `Orders` tab:
image

**Comment by lilitkarapetyan** [comment]
> I'm not sure how this problem arose, but we have a layout issue with the `Orders` tab: image

Ok, I will look into it

**Comment by IanMayo** [comment]
Hello @lilitkarapetyan - I like this piece of work. Thanks for refactoring out those helpers.

`json-editor` has a `cacheMessage` prop.  As far as I can see, that is never true.  If I'm right about that, we never call `expiredStorage.setItem` in `json-editor\index.tsx`.  If that's right, we never need to call `expiredStorage.getItem` or `expiredStorege.removeItem` in `json-editor\index.tsx`.   If that's right, we don't need to refer to `expiredStorage` in `json-editor\index.tsx` at all - and we can remove it.

Yes, there are a few _if I'm right_'s above.  Would you mind checking it please?

**Comment by lilitkarapetyan** [comment]
> 



> Hello @lilitkarapetyan - I like this piece of work. Thanks for refactoring out those helpers.
> 
> `json-editor` has a `cacheMessage` prop. As far as I can see, that is never true. If I'm right about that, we never call `expiredStorage.setItem` in `json-editor\index.tsx`. If that's right, we never need to call `expiredStorage.getItem` or `expiredStorege.removeItem` in `json-editor\index.tsx`. If that's right, we don't need to refer to `expiredStorage` in `json-editor\index.tsx` at all - and we can remove it.
> 
> Yes, there are a few _if I'm right_'s above. Would you mind checking it please?
Hello Ian , Thanks I will check it and remove `expiredStorage` from `JsonEditor`

**Comment by IanMayo** [comment]
> Thanks I will check it and remove expiredStorage from JsonEditor

The `expiredStorage` fix was a great workaround, that made a monster problem go away.  So, it feels _scary_ to delete it.  But, if we really can remove it, then great :-D

**Comment by IanMayo** [comment]
Note: this bug should be included in this PR please, @lilitkarapetyan : https://github.com/serge-web/serge/issues/2164

**Review by IanMayo — APPROVED** [review]
tested and working

**Pull request merged**